sklearn/_loss/tests/test_loss.py

```diff
@@ -1091,3 +1126,50 @@ def test_loss_pickle(loss):
     assert loss(y_true=y_true, raw_prediction=raw_prediction) == approx(
         unpickled_loss(y_true=y_true, raw_prediction=raw_prediction)
     )
+
+
+@pytest.mark.parametrize("p", [0, 1, 1.5, 2, 3])
+def test_tweedie_log_identity_consistency(p):
+    half_tweedie_log = HalfTweedieLoss(power=p)
+    half_tweedie_identity = HalfTweedieLossIdentity(power=p)
+    n_samples = 10
+    # We intentionally pick an arbitrary value for y_true and keep it fixed
+    # because the dropped constant term in the HalfTweedieLoss class depends on
+    # y_true.
+    y_true = np.full(shape=n_samples, fill_value=0.1)
+
+    # XXX: increasing the upper bound makes the test fail very easily, first
+    # with significantly different gradients, and then with different loss
+    # values.
+    raw_prediction = np.logspace(-10, -3, n_samples)
+
+    # Let's compare the loss values, up to some constant term that is dropped
+    # in HalfTweedieLoss but not in HalfTweedieLossIdentity.
+    loss_log_without_constant_terms = half_tweedie_log.loss(
+        y_true=y_true, raw_prediction=raw_prediction
+    )
+    loss_identity_with_constant_terms = half_tweedie_identity.loss(
+        y_true=y_true, raw_prediction=np.exp(raw_prediction)
+    )
+    # Check that the difference between the two ways of computing the loss
+    # values are just the constant terms that are computed by
+    # HalfTweedieLossIdentity but ignored by HalfTweedieLoss (with built-in log
+    # link)
+    diff = loss_log_without_constant_terms - loss_identity_with_constant_terms
+    assert_allclose(diff, diff[0])
+
+    # The gradient and hessian should be the same for the same because they
+    # do not depend on the constant terms.
+
+    gradient_log, hessian_log = half_tweedie_log.gradient_hessian(
+        y_true=y_true, raw_prediction=raw_prediction
+    )
+    gradient_identity, hessian_identity = half_tweedie_identity.gradient_hessian(
+        y_true=y_true, raw_prediction=np.exp(raw_prediction)
+    )
+    # XXX: is the large atol necessary to make this assertion pass hiding a bug
+    # in gradient computation?
+    assert_allclose(gradient_log, gradient_identity, atol=1e-3)
+
+    # XXX: the following always fails
+    # assert_allclose(hessian_log, hessian_identity, atol=1e-3)
```

Comment:
I updated the test in [80bf470](https://github.com/scikit-learn/scikit-learn/pull/22548/commits/80bf4702223f624ffcd66625ed8f76cb966a8c08) but the gradient and hessian computation still do not match as I would have expected. Maybe I am making a mistake in the test. I have not yet checked the source of that function to understand the discrepancy.

Reply:
It's because of the chain rule with the (log) link function, e.g. `d/dx loss(exp(x)) = exp(x) * grad(exp(x))` and so on.

The two losses must be equal if we add `constant_to_optimal_zero`, no need for `diff`.

Furthermore, we only need to check the losses. Gradients and hessians are tested in `test_gradients_hessians_numerically` (in a way that I can sleep well).

Reply:
Fixed in https://github.com/scikit-learn/scikit-learn/pull/22548/commits/225863b4d82d2f0d164fae1142ae3009a60ba295.